mlx4_ib: fix SRQ initialization per PRM requirements.

Very similar to mthca-memfree SRQ initialization.

Signed-off-by: Jack Morgenstein <jackm@dev.mellanox.co.il>

Index: ofed_kernel/drivers/infiniband/hw/mlx4/srq.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/mlx4/srq.c	2010-03-09 15:16:03.530559000 +0200
+++ ofed_kernel/drivers/infiniband/hw/mlx4/srq.c	2010-03-09 16:05:43.864875000 +0200
@@ -138,6 +138,8 @@ struct ib_srq *mlx4_ib_create_xrc_srq(st
 		if (err)
 			goto err_mtt;
 	} else {
+		struct mlx4_wqe_data_seg *scatter;
+
 		err = mlx4_db_alloc(dev->dev, &srq->db, 0);
 		if (err)
 			goto err_srq;
@@ -157,6 +159,11 @@ struct ib_srq *mlx4_ib_create_xrc_srq(st
 			next = get_wqe(srq, i);
 			next->next_wqe_index =
 				cpu_to_be16((i + 1) & (srq->msrq.max - 1));
+
+			for (scatter = (void *) (next + 1);
+			     (void *) scatter < (void *) next + desc_size;
+			     ++scatter)
+				scatter->lkey = cpu_to_be32(MLX4_INVALID_LKEY);
 		}
 
 		err = mlx4_mtt_init(dev->dev, srq->buf.npages, srq->buf.page_shift,
